File: 1039-minimum-score-triangulation-of-polygon/1039-minimum-score-triangulation-of-polygon.java
class Solution {
    int [][]dp;
    public int memo(int[]values,int stVertex,int endVertex){
        if(stVertex+1==endVertex)return 0;
        int ans=Integer.MAX_VALUE;
        if(dp[stVertex][endVertex]!=-1)return dp[stVertex][endVertex];
        for(int k=stVertex+1;k<endVertex;k++){
            ans=Math.min(ans,memo(values,stVertex,k)+memo(values,k,endVertex)+values[stVertex]*values[endVertex]*values[k]);
        }
        return dp[stVertex][endVertex]=ans;
    }
    public int minScoreTriangulation(int[] values) {
        int n=values.length;
        dp=new int[n][n];
        for(int[]a:dp)
            Arrays.fill(a,-1);
        return memo(values,0,values.length-1);
    }
}

File: 1039-minimum-score-triangulation-of-polygon/1039-minimum-score-triangulation-of-polygon.cpp
class Solution {
public:
    vector<vector<int>>dp;
    int recur(vector<int>&values,int i,int j){
        if(i+1==j)return 0;        
        if(dp[i][j]!=-1)return dp[i][j];
        int ans=INT_MAX;
        for(int k=i+1;k<j;k++){
            ans=min(ans,recur(values,i,k) + recur(values,k,j)+values[i]*values[j]*values[k]);
        }
        return dp[i][j]=ans;
    }
    
    int minScoreTriangulation(vector<int>& values) {
        dp.resize(values.size(),vector<int>(values.size(),-1));
        return recur(values,0,values.size()-1);
    }
};